pkg/services/ngalert/api/api_ruler.go

```diff
@@ -243,62 +249,73 @@ func (srv RulerSrv) RoutePostNameRulesConfig(c *models.ReqContext, ruleGroupConf
 		return toNamespaceErrorResponse(err)
 	}
 
-	//TODO: Should this belong in alerting-api?
-	if ruleGroupConfig.Name == "" {
-		return ErrResp(http.StatusBadRequest, errors.New("rule group name is not valid"), "")
+	rules, err := validateRuleGroup(&ruleGroupConfig, c.SignedInUser.OrgId, namespace, srv.BaseInterval, conditionValidator(c, srv.DatasourceCache))
+	if err != nil {
+		return ErrResp(http.StatusBadRequest, err, "")
 	}
 
-	alertRuleUIDs := make(map[string]struct{})
-	for _, r := range ruleGroupConfig.Rules {
-		cond := ngmodels.Condition{
-			Condition: r.GrafanaManagedAlert.Condition,
-			OrgID:     c.SignedInUser.OrgId,
-			Data:      r.GrafanaManagedAlert.Data,
+	return srv.updateAlertRulesInGroup(c, namespace, ruleGroupConfig, rules)
+}
+
+func (srv RulerSrv) updateAlertRulesInGroup(c *models.ReqContext, namespace *models.Folder, ruleGroupConfig apimodels.PostableRuleGroupConfig, rules []*ngmodels.AlertRule) response.Response {
+	// TODO add create rules authz logic
+
+	var changes *RuleChanges = nil
+	err := srv.store.InTransaction(c.Req.Context(), func(tranCtx context.Context) error {
+		var err error
+		changes, err = calculateChanges(tranCtx, srv.store, c.SignedInUser.OrgId, namespace, ruleGroupConfig.Name, rules)
+		if err != nil {
+			return err
 		}
-		if err := validateCondition(c.Req.Context(), cond, c.SignedInUser, c.SkipCache, srv.DatasourceCache); err != nil {
-			return ErrResp(http.StatusBadRequest, err, "failed to validate alert rule %q", r.GrafanaManagedAlert.Title)
+
+		// TODO add update/delete authz logic
+		err = srv.store.UpsertAlertRules(tranCtx, changes.Upsert)
+		if err != nil {
+			return fmt.Errorf("failed to add or update rules: %w", err)
 		}
-		if r.GrafanaManagedAlert.UID != "" {
-			_, ok := alertRuleUIDs[r.GrafanaManagedAlert.UID]
-			if ok {
-				return ErrResp(http.StatusBadRequest, fmt.Errorf("conflicting UID %q found", r.GrafanaManagedAlert.UID), "failed to validate alert rule %q", r.GrafanaManagedAlert.Title)
+
+		for _, rule := range changes.Delete {
+			if err = srv.store.DeleteAlertRuleByUID(tranCtx, c.SignedInUser.OrgId, rule.UID); err != nil {
+				return fmt.Errorf("failed to delete rule %d with UID %s: %w", rule.ID, rule.UID, err)
 			}
-			alertRuleUIDs[r.GrafanaManagedAlert.UID] = struct{}{}
 		}
-	}
 
-	numOfNewRules := len(ruleGroupConfig.Rules) - len(alertRuleUIDs)
-	if numOfNewRules > 0 {
-		// quotas are checked in advanced
-		// that is acceptable under the assumption that there will be only one alert rule under the rule group
-		// alternatively we should check the quotas after the rule group update
-		// and rollback the transaction in case of violation
-		limitReached, err := srv.QuotaService.QuotaReached(c, "alert_rule")
-		if err != nil {
-			return ErrResp(http.StatusInternalServerError, err, "failed to get quota")
-		}
-		if limitReached {
-			return ErrResp(http.StatusForbidden, errors.New("quota reached"), "")
+		if len(changes.Upsert) > 0 {
```

Comment:
any reason that you don't check `changes.newRules > 0` instead? if all rules in `changes.Upsert` are updates (no new rules) then I think there is no need to check the quota.

Reply:
fixed.